File: Binary_Search/74_search_a_2d_matrix/74.cpp


class Solution {
public:

    bool recur(vector<vector<int>> &matrix, int left, int right, int target) {
        
        if (left > right) {
            return false;
        }

        int mid = (int)((left + right) / 2);

        int col = mid % matrix[0].size();
        int row = (int)(mid / matrix[0].size());

        if (matrix[row][col] == target) {
            return true;
        }

        if (matrix[row][col] > target) {
            return recur(matrix, left, mid - 1, target);
        }
        return recur(matrix, mid + 1, right, target);
    }

    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        return recur(matrix, 0, (matrix.size() * matrix[0].size()) - 1, target);
    }
};

File: Binary_Search/74_search_a_2d_matrix/74.py


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        rows = len(matrix)
        cols = len(matrix[0])

        def recur(left, right):
            
            if left > right:
                return False
            
            mid = (left + right) // 2

            col = mid % cols
            row = mid // cols 

            if matrix[row][col] == target:
                return True
            
            if matrix[row][col] > target:
                return recur(left, mid - 1)
            return recur(mid + 1, right)
        
        return recur(0, (rows * cols) - 1)